dotnet/src/SemanticKernel/Planning/Plan.cs

```
@@ -64,45 +52,141 @@ public Task<Plan> RunNextStepAsync(IKernel kernel, ContextVariables variables, C
     [JsonPropertyName("is_semantic")]
-    public bool IsSemantic { get; } = false;
+    public bool IsSemantic { get; protected set; } = false;
 
     /// <inheritdoc/>
     [JsonPropertyName("request_settings")]
-    public CompleteRequestSettings RequestSettings { get; } = new();
+    public CompleteRequestSettings RequestSettings { get; protected set; } = new();
+
+    #endregion ISKFunction implementation
+
+    public Plan(string goal)
+    {
+        this.Description = goal;
+    }
+
+    public Plan(ISKFunction function)
+    {
+        this.SetFunction(function);
+    }
+
+    /// <summary>
+    /// Run the next step of the plan
+    /// </summary>
+    /// <param name="kernel">Kernel instance to use</param>
+    /// <param name="variables">Variables to use</param>
+    /// <param name="cancellationToken">Cancellation token</param>
+    /// <returns>The updated plan</returns>
+    public Task<Plan> RunNextStepAsync(IKernel kernel, ContextVariables variables, CancellationToken cancellationToken = default)
+    {
+        var context = new SKContext(
+            variables,
+            kernel.Memory,
+            kernel.Skills,
+            kernel.Log,
+            cancellationToken);
+        return this.InvokeNextStepAsync(context);
+    }
+
+    #region ISKFunction implementation
 
     /// <inheritdoc/>
     public FunctionView Describe()
     {
-        throw new NotImplementedException();
+        return this.Function?.Describe() ?? new();
     }
 
     /// <inheritdoc/>
-    public Task<SKContext> InvokeAsync(string input, SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
+    public virtual Task<SKContext> InvokeAsync(string input, SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
     {
-        throw new NotImplementedException();
+        if (this.Function is not null)
+        {
```

Should we throw here if steps are set? That is likely an invalid state.